# HG changeset patch
# Parent 6549ec5cf5e1cb98c5f66e3642a871378ad8770c

diff --git a/layout/generic/nsSimplePageSequence.cpp b/layout/generic/nsSimplePageSequence.cpp
--- a/layout/generic/nsSimplePageSequence.cpp
+++ b/layout/generic/nsSimplePageSequence.cpp
@@ -673,19 +673,33 @@ nsSimplePageSequenceFrame::PrePrintNextP
       return NS_OK;
     } else if (canvas->IsPrintCallbackDone()) {
       doneCounter++;
     }
   }
 
   // If all canvas have finished rendering, return true, otherwise false.
   *aDone = doneCounter == mCurrentCanvasList.Length();
+
   return NS_OK;
 }
 
+void
+nsSimplePageSequenceFrame::InvalidateInternal(const nsRect& aDamageRect,
+                                  nscoord aX, nscoord aY, nsIFrame* aForChild,
+                                  PRUint32 aFlags)
+{
+  printf("=== nsSimplePageSequenceFrame:;InvalidateInternal\n");
+  // xxx Invalidate the entire frame as otherwise invalidate of printCanvas
+  // don't work properly. This is hopefully no longer necessary once 539356
+  // lands.
+  nsContainerFrame::InvalidateInternal(
+      nsRect(nsPoint(0,0), GetSize()), 0, 0, aForChild, aFlags); 
+}
+
 NS_IMETHODIMP
 nsSimplePageSequenceFrame::ResetPrintCanvasList()
 {
   for (PRInt32 i = mCurrentCanvasList.Length() - 1; i >= 0 ; i--) {
     nsHTMLCanvasElement *canvas = mCurrentCanvasList[i];
     nsICanvasRenderingContextInternal *ctx = canvas->GetContextAtIndex(0);
 
     if (ctx)
diff --git a/layout/generic/nsSimplePageSequence.h b/layout/generic/nsSimplePageSequence.h
--- a/layout/generic/nsSimplePageSequence.h
+++ b/layout/generic/nsSimplePageSequence.h
@@ -93,16 +93,19 @@ public:
 
   /**
    * Get the "type" of the frame
    *
    * @see nsGkAtoms::sequenceFrame
    */
   virtual nsIAtom* GetType() const;
   
+  virtual void InvalidateInternal(const nsRect& aDamageRect,
+                                   nscoord aX, nscoord aY, nsIFrame* aForChild,
+                                   PRUint32 aFlags);
 #ifdef NS_DEBUG
   NS_IMETHOD  GetFrameName(nsAString& aResult) const;
 #endif
 
   void PaintPageSequence(nsRenderingContext& aRenderingContext,
                          const nsRect&        aDirtyRect,
                          nsPoint              aPt);
 
