# HG changeset patch
# Parent bbf46c67d4d37106895789add871306fbabebf3f
diff --git a/layout/printing/nsPrintEngine.cpp b/layout/printing/nsPrintEngine.cpp
--- a/layout/printing/nsPrintEngine.cpp
+++ b/layout/printing/nsPrintEngine.cpp
@@ -1661,18 +1661,19 @@ nsPrintEngine::SetupToPrintContent()
 
   // for (int i=mPrt->mPrintDocList.Length() - 1;i>=0;i--) {
-  for (PRUint32 i=1;i<mPrt->mPrintDocList.Length();i++) {
+  for (PRUint32 i=0;i<mPrt->mPrintDocList.Length();i++) {
     nsPrintObject* po = mPrt->mPrintDocList.ElementAt(i);
     NS_ASSERTION(po, "nsPrintObject can't be null!");
     printf("nsPrintEngine::SetupToPrintContent - Reconstruct %d - %p \n", i, po->mPresShell.get());
     
+    nsSize adjSize = po->mPresContext->GetPageSize();
     po->mPresShell->ReconstructFrames();
-    // Process the reflow event InitialReflow posted
-    // po->mPresShell->FlushPendingNotifications(Flush_Layout);
+    po->mPresShell->ResizeReflow(adjSize.width, adjSize.height);
+
+    printf("nsPrintEngine::SetupToPrintContent - After Reflow %d - %p \n", i, po->mPresShell.get());
+    po->mPresShell->ReconstructFrames();
   }
 
   printf("nsPrintEngine::SetupToPrintContent - EndReflow\n");
 
-  return NS_OK;
-
   // Here is where we do the extra reflow for shrinking the content
   // But skip this step if we are in PrintPreview
