diff --git a/layout/printing/nsPrintEngine.cpp b/layout/printing/nsPrintEngine.cpp
--- a/layout/printing/nsPrintEngine.cpp
+++ b/layout/printing/nsPrintEngine.cpp
@@ -263,5 +263,8 @@ nsPrintEngine::nsPrintEngine() :
   mOldPrtPreview(nsnull),
   mDebugFile(nsnull),
-  mLoadCounter(0)
+  mLoadCounter(0),
+  mSetupTimer(false),
+  mLoadTimer(nsnull),
+  mLoadedContent(false)
 {
 }
@@ -294,4 +297,5 @@ void nsPrintEngine::Destroy()
 
 #endif
+  mLoadTimer = nsnull;
   mDocViewerPrint = nsnull;
 }
@@ -321,4 +325,6 @@ nsresult nsPrintEngine::Initialize(nsIDo
   NS_ENSURE_ARG_POINTER(aDocument);
 
+  mSetupTimer = false;
+
   mDocViewerPrint = aDocViewerPrint;
   mContainer      = aContainer;
@@ -1661,5 +1667,5 @@ nsPrintEngine::SetupToPrintContent()
   }
 
-  printf("nsPrintEngine::SetupToPrintContent - BeginReflow\n");
+  printf("nsPrintEngine::SetupToPrintContent - BeginReflow %p\n", this);
 
   // Here we reflow all the PrintObjects
@@ -1883,4 +1889,10 @@ nsPrintEngine::OnStateChange(nsIWebProgr
                                nsresult aStatus)
 {
+  printf("nsPrintEngine::OnStateChange() %p\n", this);
+  if (mSetupTimer == true) {
+    printf("mLoadTimer already scheduled - reset it!\n");
+    mLoadTimer->Cancel();
+    mLoadTimer = nsnull;
+  }  
   // For some reasons, this never gets called???
   if (aStatus & nsIWebProgressListener::STATE_IS_NETWORK) {
@@ -1889,6 +1901,7 @@ nsPrintEngine::OnStateChange(nsIWebProgr
 
   if (aStateFlags & STATE_START) {
+    printf("nsPrintEngine - startRequest\n");
     mLoadCounter++;
-    printf("nsPrintEngine - startRequest\n");
+    mLoadedContent = true;
   } else if (aStateFlags & STATE_STOP) {
     printf("nsPrintEngine - stopRequest %s\n", (aStatus & nsIWebProgressListener::STATE_IS_NETWORK ? "net" : "oth"));
@@ -1898,5 +1911,5 @@ nsPrintEngine::OnStateChange(nsIWebProgr
     // are still no new requests, then another reflow.
     if (mLoadCounter == 0) {
-
+      // StartLoadTimer();
     }
   }
diff --git a/layout/printing/nsPrintEngine.h b/layout/printing/nsPrintEngine.h
--- a/layout/printing/nsPrintEngine.h
+++ b/layout/printing/nsPrintEngine.h
@@ -56,4 +56,7 @@
 #include "nsWeakReference.h"
 
+// Timer Includes
+#include "nsITimer.h"
+
 // Classes
 class nsPagePrintTimer;
@@ -309,5 +312,9 @@ protected:
   FILE* mDebugFile;
 
-  PRInt32                 mLoadCounter;
+  void                    StartLoadTimer();
+  PRInt32                     mLoadCounter;
+  bool                        mLoadedContent;
+  nsCOMPtr<nsITimer>          mLoadTimer;
+  bool                        mSetupTimer; 
 
 private:
