# HG changeset patch
# Parent b762e5916d36944bd6fc100437bd66cb20699caf

diff --git a/layout/printing/nsPrintEngine.cpp b/layout/printing/nsPrintEngine.cpp
--- a/layout/printing/nsPrintEngine.cpp
+++ b/layout/printing/nsPrintEngine.cpp
@@ -2443,7 +2443,6 @@
     return true; // means we are done preparing the page.
   }
 
-   
   // Check setting to see if someone request it be cancelled
   bool isCancelled = false;
   mPrt->mPrintSettings->GetIsCancelled(&isCancelled);
@@ -2453,7 +2452,12 @@
   // Ask mPageSeqFrame if the page is ready to be printed.
   // If the page doesn't get printed at all, the function will return true.
   bool done = false;
-  mPageSeqFrame->PrePrintNextPage(&done);
+  nsresult rv = mPageSeqFrame->PrePrintNextPage(&done);
+  if (NS_FAILED(rv)) {
+    ShowPrintErrorDialog(rv);
+    mPrt->mIsAborted = true;
+    return true;
+  }
   return done;
 }
 
@@ -2478,7 +2482,7 @@
   // Check setting to see if someone request it be cancelled
   bool isCancelled = false;
   mPrt->mPrintSettings->GetIsCancelled(&isCancelled);
-  if (isCancelled)
+  if (isCancelled || mPrt->mIsAborted)
     return true;
 
   PRInt32 pageNum, numPages, endPage;
