# HG changeset patch
# Parent 9e477b7ff635d36b33cc90a2c155fdd97b297018

diff --git a/content/canvas/src/nsCanvasRenderingContext2D.cpp b/content/canvas/src/nsCanvasRenderingContext2D.cpp
--- a/content/canvas/src/nsCanvasRenderingContext2D.cpp
+++ b/content/canvas/src/nsCanvasRenderingContext2D.cpp
@@ -1030,18 +1030,20 @@ nsCanvasRenderingContext2D::ApplyStyle(S
         color.a *= CurrentState().globalAlpha;
 
     mThebes->SetColor(color);
 }
 
 nsresult
 nsCanvasRenderingContext2D::Redraw()
 {
+    /*
     if (mIsEntireFrameInvalid)
         return NS_OK;
+    */
     mIsEntireFrameInvalid = true;
 
     if (!mCanvasElement) {
         NS_ASSERTION(mDocShell, "Redraw with no canvas element or docshell!");
         return NS_OK;
     }
 
     nsSVGEffects::InvalidateDirectRenderingObservers(HTMLCanvasElement());
@@ -1050,19 +1052,21 @@ nsCanvasRenderingContext2D::Redraw()
 
     return NS_OK;
 }
 
 NS_IMETHODIMP
 nsCanvasRenderingContext2D::Redraw(const gfxRect& r)
 {
     ++mInvalidateCount;
-
+  
+    /*
     if (mIsEntireFrameInvalid)
         return NS_OK;
+    */
 
     if (mPredictManyRedrawCalls ||
         mInvalidateCount > kCanvasMaxInvalidateCount) {
         return Redraw();
     }
 
     if (!mCanvasElement) {
         NS_ASSERTION(mDocShell, "Redraw with no canvas element or docshell!");
@@ -1074,20 +1078,22 @@ nsCanvasRenderingContext2D::Redraw(const
     HTMLCanvasElement()->InvalidateCanvasContent(&r);
 
     return NS_OK;
 }
 
 NS_IMETHODIMP
 nsCanvasRenderingContext2D::RedrawUser(const gfxRect& r)
 {
+    /*
     if (mIsEntireFrameInvalid) {
         ++mInvalidateCount;
         return NS_OK;
     }
+    */
 
     return Redraw(mThebes->UserToDevice(r));
 }
 
 NS_IMETHODIMP
 nsCanvasRenderingContext2D::SetDimensions(PRInt32 width, PRInt32 height)
 {
     Initialize(NULL, width, height);
diff --git a/content/html/content/src/nsHTMLCanvasElement.cpp b/content/html/content/src/nsHTMLCanvasElement.cpp
--- a/content/html/content/src/nsHTMLCanvasElement.cpp
+++ b/content/html/content/src/nsHTMLCanvasElement.cpp
@@ -875,21 +875,24 @@ void
 nsHTMLCanvasElement::SetWriteOnly()
 {
   mWriteOnly = true;
 }
 
 void
 nsHTMLCanvasElement::InvalidateCanvasContent(const gfxRect* damageRect)
 {
+  printf("nsHTMLCanvasElement:;InvalidateCanvasContent - Entry\n");
   // We don't need to flush anything here; if there's no frame or if
   // we plan to reframe we don't need to invalidate it anyway.
   nsIFrame *frame = GetPrimaryFrame();
-  if (!frame)
+  if (!frame) {
+    printf("nsHTMLCanvasElement:;InvalidateCanvasContent - Early Quit\n");
     return;
+  }
 
   frame->MarkLayersActive(nsChangeHint(0));
 
   nsRect invalRect;
   nsRect contentArea = frame->GetContentRect();
   if (damageRect) {
     nsIntSize size = GetWidthHeight();
     if (size.width != 0 && size.height != 0) {
@@ -908,16 +911,20 @@ nsHTMLCanvasElement::InvalidateCanvasCon
 
       invalRect = invalRect.Intersect(nsRect(nsPoint(0,0), contentArea.Size()));
     }
   } else {
     invalRect = nsRect(nsPoint(0, 0), contentArea.Size());
   }
   invalRect.MoveBy(contentArea.TopLeft() - frame->GetPosition());
 
+  printf("nsHTMLCanvasElement:;InvalidateCanvasContent - Invalidate\n");
+
+  printf("rect: %d,%d,%d,%d\n", invalRect.x, invalRect.y, invalRect.width, invalRect.height);
+
   Layer* layer = frame->InvalidateLayer(invalRect, nsDisplayItem::TYPE_CANVAS);
   if (layer) {
     static_cast<CanvasLayer*>(layer)->Updated();
   }
 }
 
 void
 nsHTMLCanvasElement::InvalidateCanvas()
