# HG changeset patch
# Parent e8560319b36e6dd537f583b4ee75d66d77d280fc
diff --git a/content/html/content/src/nsHTMLCanvasElement.cpp b/content/html/content/src/nsHTMLCanvasElement.cpp
--- a/content/html/content/src/nsHTMLCanvasElement.cpp
+++ b/content/html/content/src/nsHTMLCanvasElement.cpp
@@ -84,7 +84,7 @@
   {
   }
 
-  NS_IMETHOD GetContext(nsISupports **context)
+  NS_SCRIPTABLE NS_IMETHOD GetContext(nsISupports **context)
   {
     printf("=== GetContext\n");
     NS_ADDREF(*context = mContext);
@@ -110,18 +110,22 @@
   /* additional members */
 };
 
+NS_IMPL_ISUPPORTS1(nsCanvasPrintState, nsICanvasPrintState)
+
 NS_DEFINE_STATIC_IID_ACCESSOR(nsCanvasPrintState, NS_ICANVASPRINTSTATE_IID) 
 
-NS_IMPL_ADDREF(nsCanvasPrintState)
-NS_IMPL_RELEASE(nsCanvasPrintState)
+// NS_IMPL_ADDREF(nsCanvasPrintState)
+// NS_IMPL_RELEASE(nsCanvasPrintState)
 
 DOMCI_DATA(CanvasPrintState, nsCanvasPrintState)
 
+/*
 NS_INTERFACE_MAP_BEGIN(nsCanvasPrintState)
   NS_INTERFACE_MAP_ENTRY(nsCanvasPrintState)
   NS_INTERFACE_MAP_ENTRY(nsICanvasPrintState)
   NS_INTERFACE_MAP_ENTRY(nsISupports)
 NS_INTERFACE_MAP_END
+*/
 
 // ---------------------------------------------------------------------------
 
