# HG changeset patch
# Parent 09da1284cafe652a03f74c44b197c56a12d1192c

diff --git a/content/html/content/src/nsHTMLCanvasElement.cpp b/content/html/content/src/nsHTMLCanvasElement.cpp
--- a/content/html/content/src/nsHTMLCanvasElement.cpp
+++ b/content/html/content/src/nsHTMLCanvasElement.cpp
@@ -219,28 +219,25 @@
 bool
 nsHTMLCanvasElement::HandlePrintCallback(nsIntSize &size)
 {
-  if (!mIsPrintCanvas)
+  // If the canvas is a printCanvas and the printing target is real paper,
+  // then the canvas is setup in nsSimplePageSequence::PrePagePrintNext().
+  if (!mIsPrintCanvas || !mIsPrintPreview)
     return true;
 
   nsresult rv;
 
   // For print preview, resize the canvas context, such that
-  if (mIsPrintPreview) {
-    GetClientWidth(&size.width);
-    GetClientHeight(&size.height);
-
-    if (!mIsReadyToPrint) {
-      rv = mCurrentContext->AdjustToSize(size.width, size.height);
-
-      if (!NS_SUCCEEDED(rv))
-        return false;
-    }
-  }
+  GetClientWidth(&size.width);
+  GetClientHeight(&size.height);
 
   if (!mIsReadyToPrint) {
+    rv = mCurrentContext->AdjustToSize(size.width, size.height);
+
+    if (!NS_SUCCEEDED(rv))
+      return false;
+
     mIsReadyToPrint = true;
     CallPrintCallback();
-    // mPrintCallback->Render(mCurrentContext);
   }
   return true;
 }
